ENH: allow using Arrow writing in pyogrio.write_dataframe (use_arrow=True option)

jorisvandenbossche opened the pull request:
Follow-up on https://github.com/geopandas/pyogrio/pull/346 to expose the Arrow-based writing through geopandas-based `write_dataframe` as well.

Updated our geopandas writing tests to be parametrized on `use_arrow` for writing as well. This didn't catch any issues with the `write_arrow` itself, but we needed to do some pre-processing to handle the promotion to multi types consistently (as in the normal write path, this happens per geometry inside `ogr_write`)

There are a few things that don't work with Arrow:
- No support (yet) for writing without a geometry column
- ~~No support (yet) for writing an all-missing geometry column (the conversion from pandas to Arrow then doesn't result in the correct type, something that could be fixed in the pre-processing)~~ -> add that in the preprocessing
- No support for mixed type columns (that's an inherent limitation of Arrow, although in theory we could cast all values to string first, as that is also what the non-Arrow path eventually does when passing the values to GDAL)

brendan-ward commented on pyogrio/geopandas.py:
```diff
@@ -552,6 +565,61 @@ def write_dataframe(
+            # writing non-geometries to the geometry column
+            df = pd.DataFrame(df, copy=False)
+            df[geometry_column] = geometry
+        table = pa.Table.from_pandas(df)
```
```suggestion
        table = pa.Table.from_pandas(df, preserve_index=False)
```

(per the errors on CI)

jorisvandenbossche replied:
I would prefer to keep the default behaviour of pyarrow (the CI failures have been fixed, that due to 1) older versions of pandas not using a RangeIndex in certain cases, and 2) filtering rows conditionally for certain parametrization, again causing the index to become not a RangeIndex).

The default behaviour of pyarrow is: "The default of None will store the index as a column, except for RangeIndex which is stored as metadata only" 

Setting it to False would drop the index

jorisvandenbossche replied:
Ah, but we drop the index for the normal path as well ..

```
In [7]: df = geopandas.GeoDataFrame({"col": [1, 2], "geometry": [shapely.Point(1, 1), shapely.Point(2, 2)]}, index = pd.Index(["a", "b"], name="ids"))

In [8]: df
Out[8]: 
     col                 geometry
ids                              
a      1  POINT (1.00000 1.00000)
b      2  POINT (2.00000 2.00000)

In [9]: pyogrio.write_dataframe(df, "test_index.gpkg")
/home/joris/scipy/repos/pyogrio/pyogrio/raw.py:585: UserWarning: 'crs' was not provided.  The output dataset will not have projection information defined and may not be usable in other systems.
  warnings.warn(

In [10]: pyogrio.read_dataframe("test_index.gpkg")
Out[10]: 
   col                 geometry
0    1  POINT (1.00000 1.00000)
1    2  POINT (2.00000 2.00000)

```

brendan-ward commented on pyogrio/tests/test_geopandas_io.py:
```diff
@@ -223,15 +223,17 @@ def test_read_datetime_tz(test_datetime_tz, tmp_path):
-    df_read = read_dataframe(fpath)
+    write_dataframe(df, fpath, use_arrow=use_arrow)
+    df_read = read_dataframe(fpath, use_arrow=use_arrow)
+    if use_arrow:
```
Can you please add a brief comment explaining this?

brendan-ward commented on pyogrio/tests/test_geopandas_io.py:
```diff
@@ -1010,67 +1020,96 @@ def test_write_read_empty_dataframe_unsupported(tmp_path, ext):
 
+    if use_arrow and ext.startswith(".geojson"):
+        request.node.add_marker(
+            pytest.mark.xfail(reason="Bugs with append when writing Arrow to GeoJSON")
```
Is there a GDAL issue or version range to note in a comment here?

jorisvandenbossche replied:
Opened https://github.com/OSGeo/gdal/issues/9792 (will add that link in the comments here)

brendan-ward commented on pyogrio/tests/test_geopandas_io.py:
```diff
@@ -1329,15 +1394,19 @@ def test_custom_crs_io(tmpdir, naturalearth_lowres_all_ext):
+def test_write_read_mixed_column_values(request, tmp_path, use_arrow):
+    if use_arrow:
+        request.node.add_marker(
+            pytest.mark.xfail(reason="Arrow does not support mixed columns")
```
Do we have reason to believe this behavior will change in the future?  If not, I think it would be better to split the Arrow test into a separate test case that specifically checks that an exception of appropriate type is raised.

jorisvandenbossche replied:
Indeed, I don't think we should add support for this, so will split off and explicitly test it raises an error instead of using an xfail

brendan-ward commented on pyogrio/tests/test_geopandas_io.py:
```diff
@@ -1488,11 +1557,17 @@ def test_write_geometry_z_types_auto(
+        with pytest.raises(shapely.errors.GEOSException):
+            # TODO(Arrow)
+            # shapely fails parsing the WKB
+            read_dataframe(data_dir / "test_multisurface.gpkg", use_arrow=use_arrow)
```
```suggestion
            read_dataframe(data_dir / "test_multisurface.gpkg", use_arrow=True)
```

brendan-ward commented on pyogrio/tests/test_geopandas_io.py:
```diff
@@ -1488,11 +1557,17 @@ def test_write_geometry_z_types_auto(
+            # shapely fails parsing the WKB
+            read_dataframe(data_dir / "test_multisurface.gpkg", use_arrow=use_arrow)
+    else:
+        df = read_dataframe(data_dir / "test_multisurface.gpkg", use_arrow=use_arrow)
```
```suggestion
        df = read_dataframe(data_dir / "test_multisurface.gpkg")
```

(dropping `use_arrow=False` here because it makes it a little harder to follow)

brendan-ward reviewed (commented):
Thanks @jorisvandenbossche , this is a great start!

What do you think about mentioning the current limitations of writing to Arrow in the docstring for write_dataframe?  These seem important to communicate for this first version.

It would be nice to somehow avoid repeated traversals over the GeoDataFrame to standardize the geometries to their multi equivalents (under `promote_to_multi` option), but I'm not seeing anything obvious.  I wonder if this would be a good function to consider adding to shapely to do a single traversal over the geometry array and promote each type to its multi equivalent...

jorisvandenbossche commented:
> It would be nice to somehow avoid repeated traversals over the GeoDataFrame to standardize the geometries to their multi equivalents (under `promote_to_multi` option), but I'm not seeing anything obvious. I wonder if this would be a good function to consider adding to shapely to do a single traversal over the geometry array and promote each type to its multi equivalent...

Yes, this is currently lacking a proper helper in shapely, see https://github.com/shapely/shapely/issues/1617

jorisvandenbossche commented on pyogrio/tests/test_geopandas_io.py:
```diff
@@ -1488,11 +1557,17 @@ def test_write_geometry_z_types_auto(
-    df = read_dataframe(data_dir / "test_multisurface.gpkg")
+def test_read_multisurface(data_dir, use_arrow):
+    if use_arrow:
+        with pytest.raises(shapely.errors.GEOSException):
```
In this case I kept it in a single test because it's a very short one (that keeps it IMO a bit clearer what the difference between with and without arrow is, in this simple case)

brendan-ward commented on pyogrio/geopandas.py:
```diff
@@ -552,6 +568,74 @@ def write_dataframe(
+        import pyarrow as pa
+        from pyogrio.raw import write_arrow
+
+        if geometry_column is not None:
```
We should raise an exception if geometry_column is None, otherwise we get an error that doesn't clearly indicate that

```
>>> write_dataframe(pd.DataFrame(df.drop(columns=['geometry'])),'/tmp/test.dbf', use_arrow=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../pyogrio/pyogrio/geopandas.py", line 609, in write_dataframe
    geom_field = table.schema.field(geometry_column)
  File "pyarrow/types.pxi", line 2818, in pyarrow.lib.Schema.field
TypeError: Index must either be string or integer
```

brendan-ward reviewed (approved):
Thanks @jorisvandenbossche !  Overall, this looks great!

I think we should raise a custom error when there is no geometry being written, just to make that clear since it varies between the regular API and the Arrow API (for now).  But otherwise this looks ready to merge when you are ready.

jorisvandenbossche commented:
One remaining review comment is the one about writing the index or not: https://github.com/geopandas/pyogrio/pull/395#discussion_r1580972807

For consistency, I should probably drop the index here too? Although we should maybe reconsider that for the default write_dataframe code path as well, to actually write the index of the dataframe as a field, if not a RangeIndex?

For the use in geopandas itself, it doesn't matter, because it has its own logic (it will reset the index to a column to ensure it gets written)

theroggy commented:
> One remaining review comment is the one about writing the index or not: [#395 (comment)](https://github.com/geopandas/pyogrio/pull/395#discussion_r1580972807)
> 
> For consistency, I should probably drop the index here too? Although we should maybe reconsider that for the default write_dataframe code path as well, to actually write the index of the dataframe as a field, if not a RangeIndex?
> 
> For the use in geopandas itself, it doesn't matter, because it has its own logic (it will reset the index to a column to ensure it gets written)

I don't have an explicit preference either way, but it indeed should be consistent...

brendan-ward commented:
> One remaining review comment is the one about writing the index or not: https://github.com/geopandas/pyogrio/pull/395#discussion_r1580972807

my preference would be to drop it to be consistent, and revisit both of those in the future if needed, and possibly do so on the GeoPandas side rather than here.  There are a number of issues to consider if we do make it a column: how to handle name collisions, name truncation (e.g., shapefile), multi-indexes, and round-tripping as an index.

jorisvandenbossche commented:
OK, dropped the index here, and also added an explicit test for it (other tests will also fail in case the index would get preserved, but good to have an explicit test about it)

> revisit both of those in the future if needed, and possibly do so on the GeoPandas side rather than here.

Note GeoPandas already has considered this (they explicitly reset the index to a column in order to preserve it, if the index has a name and is not a RangeIndex)

brendan-ward commented:
Thanks @jorisvandenbossche !

brendan-ward commented:
@jorisvandenbossche it looks like testing the sdist failed after merging this: https://github.com/geopandas/pyogrio/actions/runs/8914593210/job/24482500369

jorisvandenbossche commented:
Will take a look, I suppose a missing marker (although a bit strange it only fails there)
